integrations/opentelemetry_ext_wsgi/README.rst

```diff
@@ -0,0 +1,47 @@
+OpenTelemetry WSGI Middleware
+=============================
+
+This library provides a WSGI middleware that can be used on any WSGI framework
+(such as Django / Flask) to track requests timing through OpenTelemetry.
+
+
+Usage (Flask)
+-------------
+
+.. code-block:: python
+
+    from flask import Flask
+    from opentelemetry_ext_wsgi import OpenTelemetryMiddleware
```

Comment:
Here goes my understanding:
1. According to [PEP8](https://www.python.org/dev/peps/pep-0008/#package-and-module-names), we don't want to have package name `opentelemetry_ext_wsgi`, instead, `opentelemetry.ext.wsgi` is preferred. Also, this gives benefit to control logging level consistently as @c24t mentioned during the meeting.
2. The namespace and folder structure should look like the following. The main reason to prefer `ext` over `integration` is that we can use the same folder to hold exporters (which technically is not integration). The main reason to prefer `ext` over `contrib` is that `ext` only consumes three letters considering we have 79 chars limit.
```
ext/opentelemetry-ext-wsgi/CHANGELOG.md
ext/opentelemetry-ext-wsgi/examples/
ext/opentelemetry-ext-wsgi/opentelemetry/ext/wsgi/__init__.py
ext/opentelemetry-ext-wsgi/setup.py
ext/opentelemetry-ext-wsgi/README.rst
ext/opentelemetry-ext-wsgi/tests/
```
3. The final package released to PyPI should be called `opentelemetry-ext-wsgi`. This follows what the majorities are doing in PyPI, gives the consistency and also makes it easier to search all OpenTelemetry Python packages (and set good example for ext packages outside this repo)

Reply:
I think @toumorokoshi also had some thoughts about package layout.

Reply:
Namespacing packages in this manner is pretty interesting, especially when releasing separate packages to pypi that share the same top-level namespace.

From the [packaging guide](https://packaging.python.org/guides/packaging-namespace-packages/) namespacing in general with package imports has caveats:
>  However, namespace packages come with several caveats and are not appropriate in all cases. A simple alternative is to use a prefix on all of your distributions such as import mynamespace_subpackage_a (you could even use import mynamespace_subpackage_a as subpackage_a to keep the import object short).

The only example that I'm aware of where packages with the same top import-level namespace (`opentelemetry`) have been shipped in multiple pypi packages that I've seen is with `opencensus`. I haven't had an opportunity to see exactly how this is handled yet but the installation story becomes pretty interesting in those cases since I believe pip will place packages into a top-level directory which has already been created by another package.

The reason for this, as I understand it has to do with collisions in `__init__.py` files and the order in which things are installed

Consider these package layouts:

```
pkg1/
├── pkg
│   └── ext
│       ├── __init__.py
│       └── pkg1
│           └── hello.py
└── setup.py

pkg2/
├── pkg
│   └── ext
│       ├── __init__.py
│       └── pkg2
│           └── hello.py
└── setup.py
```

Let's say the `__init__.py` in pkg1's pkg.ext looks like this:
```
from pkg1 import hello
```

And pkg2's `__init__.py` in pkg.ext looks like this:
```
from pkg2 import hello
```

When these 2 packages are both installed, the site-packages version of pkg looks like this:

```
pkg
└── ext
    ├── __init__.py
    ├── pkg1
    │   └── hello.py
    └── pkg2
        └── hello.py
```

There is only 1 `ext.__init__.py`, so what are the contents of that `__init__.py` file? Since opentelemetry will be accepting community contributions, if 2 community libraries use an `__init__.py` and both those packages are installed, the `__init__.py` state will not be known. This can cause import errors in customer applications.

Normally, namespaces are used from within a single package; however, when namespaces are shared across packages, as I understand it things can get a little weird in Python. I believe that for this reason, as far as I'm aware most packages that are uploaded to pypi have their own top level namespace.


For logging, the name can be specified manually so rather than using `logging.getLogger(__name__)`, one could use `logging.getLogger('opentelemetry.ext.wsgi')`

That being said, I've pushed up a version which implements the layout @reyang proposed 😄

Reply:
> There is only 1 `ext.__init__.py`, so what are the contents of that `__init__.py` file? Since opentelemetry will be accepting community contributions, if 2 community libraries use an `__init__.py` and both those packages are installed, the `__init__.py` state will not be known. This can cause import errors in customer applications.

In OpenTelemetry, we don't have `__init__.py` in the namespace:
```
opentelemetry-ext-foo/
├── opentelemetry
│   └── ext
│       └── foo
│           └── __init__.py
│           └── hello.py
└── setup.py

opentelemetry-ext-bar/
├── opentelemetry
│   └── ext
│       └── bar
│           └── __init__.py
│           └── hello.py
└── setup.py
```

In OpenCensus, namespace `__init__.py` file is using a special line `__path__ = __import__('pkgutil').extend_path(__path__, __name__)` following this [discussion](https://github.com/census-instrumentation/opencensus-python/issues/445#issue-399089780).

Reply:
@reyang I think generally it's not the OpenTelemetry project that will cause issues. The problem will only appear if somebody outside of the `OpenTelemetry` project releases an extension on pypi using the `opentelemetry.ext` namespace (if a custom extension is developed outside of the `OpenTelemetry` project). There's no enforcement on whether or not an `__init__.py` is used in their package paths.

Reply:
That's very true.
My takeaway is that evil people could publish a package called `open-telemetry` and ruin the customers by taking over `lib/site-packages/opentelemetry` folder. This is not a problem that we can solve by using or not using namespace.

Eventually I think how people would pick package is going to be based on the following facts:
1. Download number.
2. Whether the package is blessed by some big names (e.g. CNCF, OpenTelemetry).
3. GitHub stars and activity.

Reply:
The difference is that this may not be somebody trying to be evil 😛 

It could be that the package is useful, works properly / as intended. Everything is fine and then another package is installed that breaks the previously installed packages. These problems are kind of weird to debug and I suspect part of the reason cross package namespaces are not widely used in Python.  In Java / c++ it's a different story since namespacing is not tied to file structure 😄 

Anyway, this isn't anything we can't change in the long run if it becomes a problem - I'm more than happy to go with the flow, I figured it's just worth pointing this out as a potential problem area.

Reply:
Good points. Thanks for driving this conversation.